### PR TITLE
sim: add assertions on hostfs copy of structures

### PR DESCRIPTION
**Pull request by yamt**
## Summary
sim: add assertions on hostfs copy of structures

### CODE REVIEW

**Comment by yamt** [comment]
for some reasons it doesn't find config.h in case of cmake.

where is the cmake equivalent of https://github.com/apache/nuttx/blob/e1a1f7b5f4fc90f0d4708a9caab9e3bc00b5b7a9/arch/sim/src/Makefile#L412-L414 ?

**Comment by acassis** [comment]
@yamt please include a git log message explaining more about it.

**Comment by yamt** [comment]
> @yamt please include a git log message explaining more about it.

what kind of extra info do you want to see?

**Comment by acassis** [comment]
> > @yamt please include a git log message explaining more about it.
> 
> what kind of extra info do you want to see?

Just why you decided to include the assertions. Was it failing silently, etc.

File: arch/sim/src/sim/sim_checknuttxtypes.c [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+#include <sys/stat.h>
+
+/* hostfs types */
+
+#define __SIM__
+#include <nuttx/fs/hostfs.h>
+#undef __SIM__
+
+/* Here we make static assertions to ensure nuttx types (eg. struct stat)
+ * match the corresponding hostfs types. (eg. struct nuttx_stat_s)
```
Comment (acassis) [inline arch/sim/src/sim/sim_checknuttxtypes.c:40]: Is it dependent on host OS version, type (Linux, MacOS, FreeBSD, etc) or does it work same way for all OS?

**Comment by yamt** [comment]
> > > @yamt please include a git log message explaining more about it.
> > 
> > 
> > what kind of extra info do you want to see?
> 
> Just why you decided to include the assertions. Was it failing silently, etc.

done

File: arch/sim/src/sim/sim_checknuttxtypes.c (hunk 1 above)
Comment (yamt) [reply arch/sim/src/sim/sim_checknuttxtypes.c:40]: the latter.

File: arch/sim/src/sim/sim_checknuttxtypes.c [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ *
+ * TODO: check the alignment as well. (_Alignof)
+ */
+
+#define STATIC_ASSERT(c) static_assert(c, #c)
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/sim_checknuttxtypes.c:45]: why not directly use static_assert

File: arch/sim/src/sim/sim_checknuttxtypes.c [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+#define FIELD(a, b, f) \
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/sim_checknuttxtypes.c:46]: STATIC_ASSERT_FIELD

File: arch/sim/src/sim/sim_checknuttxtypes.c [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+FIELD(nuttx_stat_s, stat, st_rdev);
+FIELD(nuttx_stat_s, stat, st_size);
+FIELD(nuttx_stat_s, stat, st_atim);
+FIELD(nuttx_stat_s, stat, st_mtim);
+FIELD(nuttx_stat_s, stat, st_ctim);
+FIELD(nuttx_stat_s, stat, st_blksize);
+FIELD(nuttx_stat_s, stat, st_blocks);
+FIELD(nuttx_stat_s, stat, st_size);
+
+STATIC_ASSERT(sizeof(struct nuttx_stat_s) == sizeof(struct stat));
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/sim_checknuttxtypes.c:74]: add new macro STATIC_ASSERT_STRUCT? like this:
```
STATIC_ASSERT_STRUCT(nuttx_stat_s, stat);
```

File: arch/sim/src/sim/sim_checknuttxtypes.c [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+/* nuttx types */
+
+#include <dirent.h>
+#include <sys/statfs.h>
+#include <sys/stat.h>
+
+/* hostfs types */
+
+#define __SIM__
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/sim_checknuttxtypes.c:35]: why need define manually?

File: arch/sim/src/sim/sim_checknuttxtypes.c [hunk 6]
```diff
@@ -0,0 +1,100 @@
+/****************************************************************************
+ * arch/sim/src/sim/sim_checknuttxtypes.c
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/sim_checknuttxtypes.c:2]: should we move the check to fs/hostfs/hostfs.c?

File: arch/sim/src/sim/sim_checknuttxtypes.c (hunk 2 above)
Comment (yamt) [reply arch/sim/src/sim/sim_checknuttxtypes.c:45]: Just to save typing

File: arch/sim/src/sim/sim_checknuttxtypes.c (hunk 5 above)
Comment (yamt) [reply arch/sim/src/sim/sim_checknuttxtypes.c:35]: Because this is not HOSTSRC

File: arch/sim/src/sim/sim_checknuttxtypes.c (hunk 6 above)
Comment (yamt) [reply arch/sim/src/sim/sim_checknuttxtypes.c:2]: No, because this is very specific to sim.

File: arch/sim/src/sim/sim_checknuttxtypes.c (hunk 5 above)
Comment (xiaoxiang781216) [reply arch/sim/src/sim/sim_checknuttxtypes.c:35]: but sim_checknuttxtypes.c is added to SRCS

Comment (yamt) [reply arch/sim/src/sim/sim_checknuttxtypes.c:35]: This is a bit special because it needs to see nuttx definitions and hostfs ones.

Comment (xiaoxiang781216) [reply arch/sim/src/sim/sim_checknuttxtypes.c:35]: sorry, I misunderstand, the definition look right.

File: arch/sim/src/sim/sim_checknuttxtypes.c (hunk 6 above)
Comment (xiaoxiang781216) [reply arch/sim/src/sim/sim_checknuttxtypes.c:2]: Ok, let's change to sim_checkhostfs.c?

Comment (yamt) [reply arch/sim/src/sim/sim_checknuttxtypes.c]: renamed to sim_checkhostfstypes.c

File: arch/sim/src/sim/sim_checknuttxtypes.c (hunk 3 above)
Comment (yamt) [reply arch/sim/src/sim/sim_checknuttxtypes.c]: done

File: arch/sim/src/sim/sim_checknuttxtypes.c (hunk 4 above)
Comment (yamt) [reply arch/sim/src/sim/sim_checknuttxtypes.c]: when we assert more things for structs, maybe.

File: arch/sim/src/sim/sim_checkhostfstypes.c [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+STATIC_ASSERT(sizeof(struct nuttx_statfs_s) == sizeof(struct statfs));
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/* dummy function to suppress nxstyle and compiler warnings */
+
+void check_nuttx_types_dummy(void);
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/sim_checkhostfstypes.c:109]: remove the declaration

File: arch/sim/src/sim/sim_checkhostfstypes.c [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+void check_nuttx_types_dummy(void)
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/sim_checkhostfstypes.c:111]: sim_check_hostfs_types

**Comment by xiaoxiang781216** [comment]
@yamt please sqaush your patch into one.

File: arch/sim/src/sim/sim_checkhostfstypes.c (hunk 7 above)
Comment (yamt) [reply arch/sim/src/sim/sim_checkhostfstypes.c:109]: compilers warn without this

File: arch/sim/src/sim/sim_checkhostfstypes.c (hunk 8 above)
Comment (yamt) [reply arch/sim/src/sim/sim_checkhostfstypes.c]: done

**Comment by yamt** [comment]
> @yamt please sqaush your patch into one.

done

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**